notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+    "               \"path\" VARCHAR, \"loaded_image\" VARCHAR, \"referrer\" VARCHAR, \n",
+    "               \"referrer_host\" VARCHAR, \"server_ip\" VARCHAR, \n",
+    "               \"screen\" VARCHAR, \"window\" VARCHAR, \n",
+    "               \"session_length\" BIGINT, \"timezone\" VARCHAR, \n",
+    "               \"timezone_offset\" VARCHAR)\n",
+    ")\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"timestamp\") AS \"__time\", *\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "druid.display.run_task(sql)\n",
+    "druid.sql.wait_until_ready('kttm_nested')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c57016cb-ff17-420d-882b-174ec6fd5ab9",
+   "metadata": {},
+   "source": [
+    "As you can see, ingesting nested columns is straight forward. All you need to do is declare them as `TYPE('COMPLEX<json>')`, include the input field in the main SELECT clause ( `*` = all columns ) and you're done!\n",
+    "Take a look at the query example below where we access these nested fields as dimensions we can group by, metrics we can aggregate and filters we can apply:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cbc2f86d-f074-4d33-98a0-c6e5f44cc1ca",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "druid.display.sql(\"\"\"\n",
+    "SELECT\n",
+    "  JSON_VALUE(\"agent\", '$.browser') AS \"browser\",\n",
+    "  SUM( JSON_VALUE(\"event\", '$.layer' RETURNING BIGINT) ) AS \"sum_layers\",\n",
+    "  COUNT( DISTINCT JSON_VALUE(\"geo_ip\", '$.city') ) AS \"unique_cities\"\n",
+    "\n",
+    "FROM \"kttm_nested\"\n",
+    "\n",
+    "WHERE JSON_VALUE(\"geo_ip\", '$.continent') = 'South America'\n",
+    "\n",
+    "GROUP BY 1 \n",
+    "ORDER BY 3 DESC\n",
+    "\"\"\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a2cb243a-276b-4227-a485-f398ba078f09",
+   "metadata": {},
+   "source": [
+    "Since nested columns could have different fields from row to row or as their schema changes over time, you can inspect the fields that have been discovered during ingestion using the `JSON_PATHS` function on nested columns:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f66ef331-b119-4182-aed1-e8fa2e777650",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "druid.display.sql(\"\"\"\n",
+    "SELECT 'agent' as nested_column, STRING_AGG( DISTINCT JSON_PATHS(\"agent\"), ', ') paths FROM \"kttm_nested\"\n",
+    "UNION ALL\n",
+    "SELECT 'event', STRING_AGG( DISTINCT JSON_PATHS(\"event\"), ', ') paths FROM \"kttm_nested\"\n",
+    "UNION ALL\n",
+    "SELECT 'geo_ip', STRING_AGG( DISTINCT JSON_PATHS(\"geo_ip\"), ', ') paths FROM \"kttm_nested\"\n",
+    "\"\"\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a7a76a4b-d68a-43b3-87eb-239d9d72c042",
+   "metadata": {},
+   "source": [
+    "#### Enhancing Data at Ingestion\n",
```

Comment:
This is such an important / prominent topic I suggest you pull all the following sections out into a notebook on JOINs.

Reply:
already done in the joins notebook.

Reply:
but I think it makes sense here too